### PR TITLE
feat: Implement SHOW CREATE FLOW

### PR DESCRIPTION
**Pull request by irenjj**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
close: #3939

## What's changed and what's your intention?
Implement `SHOW CREATE FLOW`

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

### CODE REVIEW

**Comment by WenyXu** [comment]
Would you like to add some sqlness tests for `SHOW FLOW`?
BTW, It seems we need to add sqlness tests for creating flow and dropping flow cc @discord9 @waynexia

**Comment by irenjj** [comment]
> Would you like to add some sqlness tests for `SHOW FLOW`? BTW, It seems we need to add sqlness tests for creating flow and dropping flow cc @discord9 @waynexia
Sure, i'll add tests after some review feedback, because the output format of `SHOW CREATE FLOW` might change.

File: src/query/src/sql.rs [hunk 1]
```diff
@@ -590,6 +599,31 @@ pub fn show_create_table(
     Ok(Output::new_with_record_batches(records))
 }
 
+pub fn show_create_flow(obj_name: &ObjectName, flow_val: FlowInfoValue) -> Result<Output> {
```
Comment (evenyag) [inline src/query/src/sql.rs:602]: We have a `CreateFlow` statement that implements Display.
https://github.com/GreptimeTeam/greptimedb/blob/c78043d526134b6cad19151e5ffd25aaefc778fe/src/sql/src/statements/create.rs#L267-L287

You could construct that statement and then convert it to a string. This is how `SHOW CREATE TABLE` works.
https://github.com/GreptimeTeam/greptimedb/blob/main/src/query/src/sql/show_create_table.rs

File: src/query/src/sql.rs [hunk 2]
```diff
@@ -606,6 +618,54 @@ pub fn show_create_table(
...
+        .try_with_sql(flow_val.raw_sql())
+        .context(SyntaxSnafu)
+        .context(error::SqlSnafu)?;
+
+    let query = Box::new(
+        parser
+            .parse_query()
+            .context(SyntaxSnafu)
+            .context(error::SqlSnafu)?,
+    );
```
Comment (evenyag) [inline src/query/src/sql.rs:637]: We should use `ParserContext` to parse this for consistency. So we don't have to repeat the code creating the parser many times. But we might add a new method to construct the `ParserContext` first (e.g. `ParserContext::new`) and then call `ParserContext::parse_query()`.

@discord9 @WenyXu Do we have a better way to get the query statement? We store the raw sql in the flow value so we have to parse it again in show create flow.

**Review by waynexia — COMMENTED** [review]
Please add some sqlness tests following https://github.com/GreptimeTeam/greptimedb/blob/main/tests/README.md

**Comment by irenjj** [comment]
> Please add some sqlness tests following https://github.com/GreptimeTeam/greptimedb/blob/main/tests/README.md

It seems that the `create flow` statement randomly fails.

**Comment by discord9** [comment]
Sorry but currently flow only implement for standalone mode, could you move the testcase out of `common` folder for now? I will put it back when we have distributed support for flow service?

> > Please add some sqlness tests following https://github.com/GreptimeTeam/greptimedb/blob/main/tests/README.md
> 
> It seems that the `create flow` statement randomly fails.

File: src/sql/src/parser.rs [hunk 3]
```diff
@@ -38,6 +38,21 @@ pub struct ParserContext<'a> {
...
+        Ok(ParserContext { parser, sql })
+    }
+
+    /// Parses parser context to Query.
+    pub fn parser_query(&mut self) -> Result<Box<Query>> {
+        Ok(Box::new(self.parser.parse_query().context(SyntaxSnafu)?))
+    }
+
     /// Parses SQL with given dialect
     pub fn create_with_dialect(
```
Comment (evenyag) [inline src/sql/src/parser.rs:57]: `create_with_dialect` could reuse `ParserContext::new()` now.

File: src/sql/src/statements/show.rs [hunk 4]
```diff
@@ -132,6 +132,19 @@ impl Display for ShowCreateTable {
...
+/// SQL structure for `SHOW CREATE FLOW`.
+#[derive(Debug, Clone, PartialEq, Eq, Visit, VisitMut)]
+pub struct ShowCreateFlow {
+    pub flow_name: ObjectName,
+}
+
+impl Display for ShowCreateFlow {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let flow_name = &self.flow_name;
+        write!(f, r#"SHOW CREATE FLOW {flow_name}"#)
```
Comment (waynexia) [inline src/sql/src/statements/show.rs:144]: nit: `r#` looks unnecessary here

File: tests/cases/standalone/show_create_flow.result [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+CREATE FLOW filter_numbers  SINK TO out_num_cnt AS SELECT number FROM numbers_input where number > 10;
+
+Affected Rows: 0
+
+SHOW CREATE FLOW filter_numbers;
+
++----------------+---------------------------------------------------------------------------------------------------------------------------------+
+| Flow           | Create Flow                                                                                                                     |
++----------------+---------------------------------------------------------------------------------------------------------------------------------+
+| filter_numbers | CREATE OR REPLACE TASK IF NOT EXISTS filter_numbers OUTPUT AS out_num_cnt AS SELECT number FROM numbers_input WHERE number > 10 |
```
Comment (waynexia) [inline tests/cases/standalone/show_create_flow.result:25]: Could you add some line breaks like `SHOW CREATE TABLE`?

File: tests/cases/standalone/show_create_flow.result [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+Affected Rows: 0
+
+create table out_num_cnt (
+    number INT,
+    ts TIMESTAMP DEFAULT CURRENT_TIMESTAMP TIME INDEX);
+
+Affected Rows: 0
+
+CREATE FLOW filter_numbers  SINK TO out_num_cnt AS SELECT number FROM numbers_input where number > 10;
```
Comment (WenyXu) [inline tests/cases/standalone/show_create_flow.result:16]:
```suggestion
CREATE FLOW filter_numbers SINK TO out_num_cnt AS SELECT number FROM numbers_input where number > 10;
```

**Review by discord9 — APPROVED** [review]

**Comment by tisonkun** [comment]
@irenjj Thanks for your contribution! Would you share your email address to me (by email to [tison@greptime.com](mailto:tison@greptime.com))?

I'd like to send you a letter for further engagement :D

**Comment by irenjj** [comment]
> @irenjj Thanks for your contribution! Would you share your email address to me (by email to [tison@greptime.com](mailto:tison@greptime.com))?
> 
> I'd like to send you a letter for further engagement :D

Sure! I've already sent the email.

**Pull request merged**